### PR TITLE
feat(cassandra): Add multi-DC support via per-datacenter execution profiles

### LINKED ISSUE #6433: In Cassandra online store, for each feature store you can have hosts from multiple local datacenters
**Is your feature request related to a problem? Please describe.**
Because user can give only one datacenter per project, in load balancing policy, So user "should" give all hosts from same region also. 
So if they have multiple hosts from different region, they will be forced to create different project 
Also even if it connects automatically, that's not good for few reasons, one that I can think of is it can connect to "worse" regions too.. especially in production

One of the use case that I can think of, is let's say in distributed production systems , with lets say some consistency level among Cassandra nodes, let's say "one", user should be able to say that for reading this time, read from this dc and for other time read from different dc, as data might be flowing to different regions in separate time 
Even for other consistency levels , data might not be always consistent, and allowing users that power to choose - to which region to read from - based on some pipeline logic for feature engineering or materialization that user has written - is worth the trouble in my opinion.

**Describe the solution you'd like**
User should be able to decide what DataCenter they want to read from and single feature store can have multiple hosts 

**Describe alternatives you've considered**
I don' think alternative exists. 

**Additional context**

### PR DESCRIPTION
**Pull request by singhhimanshu0811**
##Which issue(s) this PR fixes
Fixes #6433 

## Summary

Adds a `datacenters` config list to `CassandraOnlineStoreConfig` as an alternative to the flat `hosts` field, enabling proper multi-datacenter support.

Each datacenter entry gets a named Cassandra **execution profile** (keyed by `local_dc`), so pipeline code can route reads/writes to a specific DC by passing `execution_profile="<dc-name>"` to the driver. The default profile is pinned to `load_balancing.local_dc`, or the first DC when `load_balancing` is absent.

### Changes
- New `CassandraDatacenterConfig` inner model with `local_dc`, `hosts`, `replication_factor` (informational), `replication_strategy` (informational)
- New `datacenters: Optional[List[CassandraDatacenterConfig]]` field on `CassandraOnlineStoreConfig`
- New Branch B in `_get_session()` that builds per-DC execution profiles and connects with all DC hosts merged; original `hosts`/`secure_bundle_path` path (Branch A) is completely untouched
- `_dc_execution_profiles: List[str]` attribute exposes registered profile names
- Updated `docs/reference/online-stores/cassandra.md` with a multi-DC `feature_store.yaml` example

### Motivation

The flat `hosts` + single `local_dc` model forces all hosts to belong to the same datacenter. Users with multi-DC clusters had to create separate feature store projects per DC. With this change:
- A single feature store config can span multiple DCs
- The Cassandra driver routes to the correct DC via execution profiles
- Per-DC `DCAwareRoundRobinPolicy` prevents reads from spilling into unintended regions
- Advanced pipelines can explicitly target a DC (e.g. read from the region where data has already been replicated)

### Backward compatibility
Fully backward compatible — `datacenters` is optional and the original code path is untouched.

### CODE REVIEW

**Review by ntkathole — COMMENTED** [review]
## Review of PR #6434 — Multi-DC Cassandra Support

Thanks for the contribution, @singhhimanshu0811! This directly addresses the gap we discussed around single-DC limitations in the Cassandra online store config. Overall the approach is solid — using per-DC execution profiles is the right driver-level mechanism. I have several items ranging from bugs to design suggestions:

---

### 1. Bug: Mutable class-level default for `_dc_execution_profiles`

```python
_dc_execution_profiles: List[str] = []
```

This is a **mutable class attribute shared across all instances**. If multiple `CassandraOnlineStore` instances are created (e.g., in tests or multi-project setups), they'll all append to the **same list**. This same pattern issue exists for `_prepared_statements` in the original code, but introducing a new one should be avoided.

**Fix**: Initialize in `_get_session` or use a property pattern:

```python
_dc_execution_profiles: List[str]

def _get_session(self, config: RepoConfig):
    ...
    if online_store_config.datacenters:
        self._dc_execution_profiles = []  # reset per connection
        ...
```

---

### 2. Bug: Missing mutual exclusivity validation

The docstring says:
> Mutually exclusive with `hosts` and `secure_bundle_path`.

But there's **no actual validation** enforcing this. If a user sets both `datacenters` and `hosts`, Branch B runs (because it checks `datacenters` first) and silently ignores `hosts`. This will confuse users.

**Fix**: Add a Pydantic `model_validator` or an explicit check at the top of `_get_session`:

```python
if online_store_config.datacenters and (
    online_store_config.hosts or online_store_config.secure_bundle_path
):
    raise CassandraInvalidConfig(
        "Cassandra config: 'datacenters' is mutually exclusive with "
        "'hosts' and 'secure_bundle_path'"
    )
```

---

### 3. Design: Profiles are created but never used by Feast operations

The PR creates named execution profiles and exposes them via `_dc_execution_profiles`, but **none of the existing Feast operations (`online_read`, `online_write_batch`, `update`) pass `execution_profile=` to the driver**. All queries will use `EXEC_PROFILE_DEFAULT` — effectively still single-DC behavior.

This means:
- The profiles are registered but dead code from Feast's perspective
- Users can't actually route reads/writes to a specific DC without modifying Feast internals

I understand this may be intended as a foundational PR with follow-up work to add `target_dc` parameters to the Feast API. If so, please mention this explicitly in the PR description and/or open a tracking issue for the follow-up. Otherwise, consider adding at minimum an optional `execution_profile` pass-through in `_read_rows_by_entity_keys` and `_write_rows_concurrently`.

---

### 4. Design: `replication_factor` and `replication_strategy` are purely informational

These fields are documented as "informational" and Feast doesn't use them. This raises questions:
- If they're never consumed by code, why are they in the config? They'll confuse users who expect Feast to act on them.
- If the intent is future keyspace creation support, consider deferring these fields to that PR to keep this one focused.

At minimum, add a log warning if these are set:

```python
if dc.replication_factor or dc.replication_strategy:
    logger.info(
        "Cassandra multi-DC: replication_factor/replication_strategy are "
        "informational only; keyspace must be pre-created."
    )
```

---

### 5. Suggestion: Duplicate `local_dc` naming is confusing

In the config structure, `local_dc` appears in two places with different semantics:
- `datacenters[*].local_dc` — the datacenter **name** (identifier)
- `load_balancing.local_dc` — which DC is the **default**

Consider renaming the per-DC field to just `name` or `dc_name` for clarity:

```yaml
datacenters:
  - name: dc1        # clearer than local_dc
    hosts: [...]
  - name: dc2
    hosts: [...]
load_balancing:
    local_dc: dc1    # this selects the default
```

The field is called `local_dc` because that's what the driver's `DCAwareRoundRobinPolicy` parameter is named, but at the config level it reads as "this datacenter's local_dc" which is semantically odd.

---

### 6. Suggestion: Empty `datacenters` list edge case

What happens if `datacenters: []` (empty list)? The truthiness check `if online_store_config.datacenters:` will be `False`, so it falls through to Branch A. But Branch A will then fail with `E_CASSANDRA_NOT_CONFIGURED` because `hosts` is also None. The error message will be misleading.

**Fix**: Validate that `datacenters` contains at least one entry if set:

```python
if online_store_config.datacenters is not None and len(online_store_config.datacenters) == 0:
    raise CassandraInvalidConfig("Cassandra 'datacenters' list must not be empty")
```

---

### 7. Tests needed

There are no unit tests for the new code path. Please add tests covering:
- Config parsing with `datacenters` field
- Mutual exclusivity validation (once added)
- Default DC selection (from `load_balancing.local_dc` vs. first-in-list fallback)
- Error case: `load_balancing.local_dc` doesn't match any DC entry
- Integration with the existing `CassandraOnlineStoreCreator` test fixture (even if it's a single-DC test, it validates backward compat)

---

### 8. Minor: Redundant `if not self._session:` in Branch A

After Branch B was inserted, the existing code still has:

```python
if not self._session:
    # configuration consistency checks
    ...
```

This `if not self._session:` is now always `True` when reached (because we already returned early at line 294 `if self._session: return`). It's harmless but misleading — the original code had it as the single branch, now it's dead logic. Consider removing the `if` wrapper for clarity (keeping the body).

---

### Summary

| Item | Severity | Action needed |
|---|---|---|
| Mutable class-level list | Bug | Must fix |
| Missing mutual exclusivity check | Bug | Must fix |
| Profiles unused by Feast ops | Design gap | Document plan or implement |
| Informational-only fields | Design | Consider removing or add warning |
| `local_dc` naming confusion | Suggestion | Consider rename |
| Empty list edge case | Suggestion | Add validation |
| No tests | Required | Add unit tests |
| Redundant `if not self._session` | Minor | Cleanup |

The core idea is good and aligns with a real production need. Looking forward to the next iteration!

**Comment by singhhimanshu0811** [comment]
Hey @ntkathole thanks for replying back. I'll revert back with updated pr in 1-2 days, meanwhile could you review these once
for point 1 , I was wondering if we could solve the prepared statement with this pr only ,
for 3, i was planning to do it later, but now that I have some time, I can resolve in this pr only. will do this. 
for 4, i ve had few questions, why is there no provision to create keyspace in the code? perhaps we can have a _create_keyspace function?
The rest, are great points, will resolve them in new push.

**Comment by singhhimanshu0811** [comment]
@ntkathole for solving issue 3, basically incorporating profiles in read and write, i am thinking something like this 

online_store:
    type: cassandra
    keyspace: feast_keyspace
    datacenters:
        - name: dc1
          hosts: [192.168.1.1, 192.168.1.2]
        - name: dc2
          hosts: [10.0.0.1]
    routing:
        read_dc: dc2    # reads go to dc2
        write_dc: dc1   # writes go to dc1
        
        
basically along with datacenters, adding a routing header, so user can decide at start time, that in this feature store session, which dc they would like to read from or which dc they would like to write from. does this work?



and of course there would be validation that routing header params are legal, i.e in datacenter header params

**Comment by ntkathole** [comment]
@singhhimanshu0811 
- I would recommend making _create_keyspace a separate PR.
- For routing config -  yes, I think that covers common use cases, - Make both read_dc and write_dc fields optional with a default value.

**Comment by ntkathole** [comment]
@singhhimanshu0811 Please fix linting here as well

**Comment by franciscojavierarceo** [comment]
@codex review

File: sdk/python/feast/infra/online_stores/cassandra_online_store/cassandra_online_store.py [hunk 1]
```diff
@@ -498,6 +733,7 @@ def _write_rows_concurrently(
             insert_cql,
             rows,
             concurrency=config.online_store.write_concurrency,
+            execution_profile=self._write_execution_profile,
```
[the remark this answers is not in the document]
Comment (singhhimanshu0811) [reply sdk/python/feast/infra/online_stores/cassandra_online_store/cassandra_online_store.py:738]: @franciscojavierarceo @ntkathole would it be fine to bump up the cassandra version?

[the remark this answers is not in the document]
Comment (franciscojavierarceo) [reply sdk/python/feast/infra/online_stores/cassandra_online_store/cassandra_online_store.py:738]: yeah

[the remark this answers is not in the document]
Comment (singhhimanshu0811) [reply sdk/python/feast/infra/online_stores/cassandra_online_store/cassandra_online_store.py:738]: @franciscojavierarceo can you kindly review again?

[the remark this answers is not in the document]
Comment (singhhimanshu0811) [reply sdk/python/feast/infra/online_stores/cassandra_online_store/cassandra_online_store.py:738]: @franciscojavierarceo can you also trigger linder tests just to confirm?

[the remark this answers is not in the document]
Comment (singhhimanshu0811) [reply sdk/python/feast/infra/online_stores/cassandra_online_store/cassandra_online_store.py:738]: upgraded versions

**Comment by singhhimanshu0811** [comment]
hi @ntkathole @franciscojavierarceo could you kindly review this pr once , could this be merged if there are no pending issues? Thanks!!Please let me know in case of some issue!

**Comment by ntkathole** [comment]
@singhhimanshu0811 Need to regenerate pixi.lock file since pyproject.toml is changed, otherwise looks good to me

**Comment by singhhimanshu0811** [comment]
Hey @ntkathole i regenarted pixi, rebased and pushed again. can you review

**Review by ntkathole — APPROVED** [review]
